GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs

```diff
@@ -224,106 +232,140 @@ internal void AddColumn(ColumnProvider columnProvider)
+            if (revision is not null && !revision.IsArtificial && _highlightAuthoredRevisions && AuthorHighlighting?.IsHighlighted(revision) != false)
             {
                 return _authoredHighlightBrush;
             }
 
-            if (rowIndex % 2 == 0 && AppSettings.RevisionGraphDrawAlternateBackColor)
+            if (_revisionGraphDrawAlternateBackColor && rowIndex % 2 == 0)
             {
                 return _alternatingRowBackgroundBrush;
             }
 
             return SystemBrushes.Window;
         }
 
+        private CellStyle? _cellStyle = null;
+        private GitRevision? _revision = null;
+        private bool _isSelected = false;
+
         private void OnCellPainting(object? sender, DataGridViewCellPaintingEventArgs e)
         {
             _lastRepaint.Restart();
 
             DebugHelpers.Assert(_rowHeight != 0, "_rowHeight != 0");
 
-            GitRevision? revision = GetRevision(e.RowIndex);
-
             if (e.RowIndex < 0 ||
                 e.RowIndex >= RowCount ||
-                !e.State.HasFlag(DataGridViewElementStates.Visible) ||
-                revision is null)
+                !e.State.HasFlag(DataGridViewElementStates.Visible))
             {
                 return;
             }
 
-            Brush backBrush = GetBackground(e.State, e.RowIndex, revision);
-            e.Graphics!.FillRectangle(backBrush, e.CellBounds);
+            GitRevision? revision = GetRevision(e.RowIndex);
+            bool isSelected = e.State.HasFlag(DataGridViewElementStates.Selected);
+            if (revision != _revision || _isSelected != isSelected)
+            {
+                _revision = revision;
+                _isSelected = isSelected;
+                if (_revision is null)
+                {
+                    _cellStyle = null;
 
-            if (Columns[e.ColumnIndex].Tag is ColumnProvider provider)
+                    return;
+                }
+
+                Brush backBrush = GetBackground(isSelected, e.RowIndex, _revision);
+                bool isNonRelativeGray = _revisionGraphDrawNonRelativesTextGray && !RowIsRelative(e.RowIndex);
+                Color foreColor = GetForeground(isSelected, isNonRelativeGray);
+                Color commitBodyForeColor = GetCommitBodyForeground(isSelected, isNonRelativeGray);
+                _cellStyle = new(backBrush, foreColor, commitBodyForeColor, _normalFont, _boldFont, _monospaceFont);
+            }
+
+            if (_cellStyle is null)
             {
-                Color foreColor = GetForeground(e.State, e.RowIndex);
-                Color commitBodyForeColor = GetCommitBodyForeground(e.State, e.RowIndex);
-                CellStyle cellStyle = new(backBrush, foreColor, commitBodyForeColor, _normalFont, _boldFont, _monospaceFont);
+                return;
+            }
+
+            e.Graphics!.FillRectangle(_cellStyle.Value.BackBrush, e.CellBounds);
+
+            if (_revision is null)
```

Comment:
The compiler disagree with you...

Reply:
> The compiler disagrees with you...

Because it does not analyze the code that far.
As we are sure, `_revision!` can be used (as in f33646202667a9fc59c57a462f751e65755e3273).